kola/tests/update/update.go

```diff
@@ -0,0 +1,172 @@
+// Copyright 2018 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package update
+
+import (
+	"bufio"
+	"fmt"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/coreos/go-omaha/omaha"
+
+	"github.com/coreos/mantle/kola"
+	"github.com/coreos/mantle/kola/cluster"
+	"github.com/coreos/mantle/kola/register"
+	tutil "github.com/coreos/mantle/kola/tests/util"
+	"github.com/coreos/mantle/platform"
+	"github.com/coreos/mantle/platform/local"
+	"github.com/coreos/mantle/util"
+)
+
+func init() {
+	register.Register(&register.Test{
+		Name:        "coreos.update.payload",
+		Run:         payload,
+		ClusterSize: 1,
+		NativeFuncs: map[string]func() error{
+			"Omaha": Serve,
+		},
+	})
+}
+
+func Serve() error {
+	omahaserver, err := omaha.NewTrivialServer(":34567")
+	if err != nil {
+		return fmt.Errorf("creating trivial omaha server: %v\n", err)
+	}
+
+	omahawrapper := local.OmahaWrapper{TrivialServer: omahaserver}
+
+	if err = omahawrapper.AddPackage("/updates/update.gz", "update.gz"); err != nil {
+		return fmt.Errorf("bad payload: %v", err)
+	}
+
+	return omahawrapper.Serve()
+}
+
+func payload(c cluster.TestCluster) {
+	addr := ConfigureOmahaServer(c, c.Machines()[0])
+
+	// create the actual test machine, the machine
+	// that is created by the test registration is
+	// used to host the omaha server
+	m, err := c.NewMachine(nil)
+	if err != nil {
+		c.Fatalf("creating test machine: %v", err)
+	}
+
+	ConfigureMachineForUpdate(c, m, addr)
+
+	tutil.AssertBootedUsr(c, m, "USR-A")
+
+	UpdateMachine(c, m)
+
+	tutil.AssertBootedUsr(c, m, "USR-B")
+
+	tutil.InvalidateUsrPartition(c, m, "USR-A")
+
+	UpdateMachine(c, m)
+
+	tutil.AssertBootedUsr(c, m, "USR-A")
+}
+
+func ConfigureOmahaServer(c cluster.TestCluster, srv platform.Machine) string {
+	if kola.UpdatePayloadFile == "" {
+		c.Skip("no update payload provided")
+	}
+
+	in, err := os.Open(kola.UpdatePayloadFile)
+	if err != nil {
+		c.Fatalf("opening update payload: %v", err)
+	}
+	defer in.Close()
+	if err := platform.InstallFile(in, srv, "/updates/update.gz"); err != nil {
+		c.Fatalf("copying update payload to omaha server: %v", err)
+	}
+
+	c.MustSSH(srv, fmt.Sprintf("sudo systemd-run --quiet ./kolet run %s Omaha", c.Name()))
```

Comment:
There should also be a verification that the server is running before the address is returned, otherwise this is racing with the launching of the new machine.

Reply:
In practice, is it likely for the kolet launch to lose the race with the instance launch?

Reply:
No. In reality we can probably leave this be, but I also think it would be nice to verify it's serving for the chance that kolet fails to launch correctly. `systemd-run` can quite easily eat errors and just spit out "yeah, I ran a unit" even if the unit instantly failed.

Reply:
`update_engine` retries even in cases of no connectivity to the omaha server so this would only really validate that the omaha server eventually starts correctly. Is the thought to detect that failure case and fail the test earlier rather than failing when the machine fails to update and deciphering the logs?

Reply:
That's a benefit, yeah. It's mostly about trying to have the function have a clear contract that it abides by.
Returning the address implies that it's already running unless there's documentation to the contrary.